### PR TITLE
Make 'ComponentId' type opaque

### PR DESCRIPTION
**Pull request by hvr**

### CODE REVIEW

**Comment by ezyang** [comment]
OK.

File: Cabal/Distribution/Package.hs [hunk 1]
```diff
@@ -175,7 +175,7 @@ instance NFData Module where
 -- @since 2.0
 
 data ComponentId
-    = ComponentId String
+    = ComponentId ShortText
```
Comment (dcoutts) [inline Cabal/Distribution/Package.hs]: @hvr I believe component IDs are actually ASCII only (indeed a small subset of ASCII), so you could use ShortByteString.

@ezyang can you confirm and tell us what the characters are so we can document that here (and perhaps add a validation function / assertion).

Comment (ezyang) [inline Cabal/Distribution/Package.hs]: Well, ComponentId valid characters coincide with what InstalledPackageId used to be: alphanumeric, dash, underscore and period.  isAlphaNum selects Unicode characters, so strictly speaking they are not ASCII. In practice they won't be ASCII if a package name has non-ASCII characters.

Comment (dcoutts) [inline Cabal/Distribution/Package.hs]: Oh yes, it goes back to the PackageName issue. Ok. ShortText is fine until we resolve that one.

**Pull request merged**